Sources/SwiftProtobuf/Message+AsyncMessageSequence.swift

```diff
@@ -0,0 +1,232 @@
+  ///           `BinaryDecodingError.tooLarge` if a size delimiter of 2GB or greater is found.
+  public func decodedBinaryDelimitedMessages<M: Message>(
+    messageType: M.Type,
+    extensions: ExtensionMap? = nil,
+    partial: Bool = false,
+    options: BinaryDecodingOptions = BinaryDecodingOptions()
+  ) -> AsyncMessageSequence<Self, M> {
+    AsyncMessageSequence<Self, M>(
+      baseSequence: self,
+      extensions: extensions,
+      partial: partial,
+      options: options
+    )
+  }
+}
+
+@available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+/// An asynchronous sequence of messages decoded from an asynchronous sequence of bytes.
+public struct AsyncMessageSequence<Base: AsyncSequence, M: Message>:
+  AsyncSequence, Sendable
+where Base: Sendable, Base.Element == UInt8 {
+  
+  /// The message type in this asynchronous sequence.
+  public typealias Element = M
+  
+  private let baseSequence: Base
+  private let extensions: ExtensionMap?
+  private let partial: Bool
+  private let options: BinaryDecodingOptions
+  
+  /// Reads size-delimited messages from the given sequence of bytes. Delimited
+  /// format allows a single file or stream to contain multiple messages. A delimited message
+  /// is a varint encoding the message size followed by a message of exactly that size.
+  ///
+  /// - Parameters:
+  ///   - baseSequence: The `AsyncSequence` to read messages from.
+  ///   - extensions: An `ExtensionMap` used to look up and decode any extensions in
+  ///    messages encoded by this sequence, or in messages nested within these messages.
+  ///   - partial: If `false` (the default),  after decoding a message, `Message.isInitialized`
+  ///     will be checked to ensure all fields are present. If any are missing,
+  ///     `BinaryEncodingError.missingRequiredFields` will be thrown.
+  ///   - options: The BinaryDecodingOptions to use.
+  /// - Returns: An asynchronous sequence of messages read from the `AsyncSequence` of bytes.
+  /// - Throws: `BinaryDecodingError` if decoding fails, throws
+  ///           `BinaryDelimited.Error` for some reading errors,
+  ///           `BinaryDecodingError.truncated` if the stream ends before fully decoding a
+  ///           message or a delimiter,
+  ///           `BinaryDecodingError.malformedProtobuf`if a delimiter could not be read and
+  ///           `BinaryDecodingError.tooLarge` if a size delimiter of 2GB or greater is found.
+  public init(
+    baseSequence: Base,
+    extensions: ExtensionMap? = nil,
+    partial: Bool = false,
+    options: BinaryDecodingOptions = BinaryDecodingOptions()
+  ) {
+    self.baseSequence = baseSequence
+    self.extensions = extensions
+    self.partial = partial
+    self.options = options
+  }
+  
+  /// An asynchronous iterator that produces the messages of this asynchronous sequence
+  @frozen public struct AsyncIterator: AsyncIteratorProtocol, Sendable
+  where Base.AsyncIterator: Sendable {
```

Comment:
This is important. We should not use `@frozen` here and async iterators must never be `Sendable`
```suggestion
  public struct AsyncIterator: AsyncIteratorProtocol {
```

Reply:
Done.